src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs

```diff
@@ -0,0 +1,488 @@
+                                                      !resultFromFlowAnalysis.SymbolsRead.Contains(localSymbol);
+
+                        var isRemovableAssignment = IsRemovableAssignmentWithoutSideEffects(unreadWriteOperation);
+
+                        if (isUnusedLocalAssignment &&
+                            !isRemovableAssignment &&
+                            _options.UnusedValueAssignmentPreference == UnusedValuePreference.UnusedLocalVariable)
+                        {
+                            // Meets current user preference of using unused local symbols for storing computation result.
+                            // Skip reporting diagnostic.
+                            return false;
+                        }
+
+                        properties = s_propertiesMap[(_options.UnusedValueAssignmentPreference, isUnusedLocalAssignment, isRemovableAssignment)];
+                        return true;
+                    }
+
+                    // Indicates if the given unused symbol write is a removable assignment.
+                    // This is true if the expression for the assigned value has no side effects.
+                    bool IsRemovableAssignmentWithoutSideEffects(IOperation unusedSymbolWriteOperation)
+                    {
+                        if (unusedSymbolWriteOperation.Parent is IAssignmentOperation assignment &&
+                            assignment.Target == unusedSymbolWriteOperation)
+                        {
+                            if (assignment.Value.ConstantValue.HasValue)
+                            {
+                                // Constant expressions have no side effects.
+                                return true;
+                            }
+
+                            switch (assignment.Value.Kind)
+                            {
+                                // Parameter/local references have no side effects and can be removed.
+                                case OperationKind.ParameterReference:
+                                case OperationKind.LocalReference:
+                                    return true;
+
+                                // Field references with null instance (static fields) or 'this' or 'Me' instance can
+                                // have no side effects and can be removed.
+                                case OperationKind.FieldReference:
+                                    var fieldReference = (IFieldReferenceOperation)assignment.Value;
+                                    return fieldReference.Instance == null || fieldReference.Instance.Kind == OperationKind.InstanceReference;
+                            }
+                        }
+                        else if (unusedSymbolWriteOperation.Parent is IIncrementOrDecrementOperation)
+                        {
+                            // Increment or decrement operations have no side effects.
```

Comment:
i on't understand this.  aren't increment/decrement ops side-effecting by definition :)

Reply:
I think my wording is not good here. What I mean is that if the new value assigned to the incremented/decremented variable is unused, then it is safe to remove the entire increment/decrement operation, as it cannot have side effects on anything but the variable. Let me improve the comment.

---
In reply to: [234868109](https://github.com/dotnet/roslyn/pull/30777#discussion_r234868109) [](ancestors = 234868109)